File: DKAudioPlayer/DKAudioPlayer.m
//
//  DKAudioPlayer.m
//
//  Created by Denis Kutlubaev on 27.02.14.

#import "DKAudioPlayer.h"

@interface DKAudioPlayer()
{
    float _playerHeight;
    float _playerWidth;
    float _inset;
}

@property (nonatomic, strong) AVAudioPlayer *audioPlayer;
@property (nonatomic, strong) NSString *durationString;
@property (nonatomic, strong) NSTimer *timer;
@property (nonatomic, strong) UILabel *timeLabel;
@property (nonatomic, strong) UIView *bubbleView;
@property (nonatomic, strong) UISlider *slider;
@property (nonatomic, strong) UIView *backgroundView;
@property (nonatomic, strong) UIButton *playPauseButton;

@end


@implementation DKAudioPlayer

#pragma mark - Initializers

- (instancetype)initWithData:(NSData *)audioData parentViewController:(UIViewController *)parentViewController
{
    // initalize audio player from nsData
    [self initAudioPlayerWithData:audioData];

    // initialize DKAudio with view controller
    self = [self initWithViewController:parentViewController];

    return self;
}

- (instancetype)initWithData:(NSData *)audioData width:(CGFloat)width height:(CGFloat)height
{
    // initalize audio player from nsData
    [self initAudioPlayerWithData:audioData];

    // initialize DK player with width and height
    self = [self initWithWidth:width height:height];

    return self;
}

- (instancetype)initWithAudioFilePath:(NSString *)audioFilePath parentViewController:(UIViewController *)parentViewController
{
    // initialize audio player from NSString
    [self initAudioPlayerWithString:audioFilePath];

    // initialize DKAudio with view controller
    self = [self initWithViewController:parentViewController];

    return self;
}

- (instancetype)initWithAudioFilePath:(NSString *)audioFilePath width:(CGFloat)width height:(CGFloat)height
{
    // initialize audio player from NSString
    [self initAudioPlayerWithString:audioFilePath];

    // initialize DK player with width and height
    self = [self initWithWidth:width height:height];

    return self;
}

- (instancetype)initWithViewController:(UIViewController *) parentViewController
{
    CGRect frame = CGRectMake(0, 0, parentViewController.view.bounds.size.width, 75.0);

    // initialize DK player with width and height
    self = [self initWithWidth:frame.size.width height:frame.size.height];

    frame.origin.y = parentViewController.view.bounds.size.height;
    self.frame = frame;
    [parentViewController.view addSubview:self];
    [parentViewController.view bringSubviewToFront:self];

    return self;
}

- (instancetype)initWithWidth:(CGFloat)width height:(CGFloat)height
{
    if (height == 0) height = 75.0;

    CGRect frame = CGRectMake(0, 0, width, height);

    self = [super initWithFrame:frame];

    if (self) {

        _playerHeight = frame.size.height;
        _playerWidth = frame.size.width;
        _inset = 15;


        self.audioPlayer.volume = 0.5;
        self.audioPlayer.delegate = self;


        [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayback error:nil];
        [[AVAudioSession sharedInstance] setActive: YES error:nil];

        long totalPlaybackTime = self.audioPlayer.duration;
        int tHours = (int)(totalPlaybackTime / 3600);
        int tMins = (int)((totalPlaybackTime/60) - tHours*60);
        int tSecs = (int)(totalPlaybackTime % 60 );
        _durationString = (tHours > 0) ? [NSString stringWithFormat:@"%i:%02d:%02d", tHours, tMins, tSecs ] : [NSString stringWithFormat:@"%02d:%02d", tMins, tSecs];

        _timer = [NSTimer scheduledTimerWithTimeInterval:0.5 target:self selector:@selector(onTimer:) userInfo:nil repeats:YES];

        self.clipsToBounds = NO;
        self.autoresizesSubviews = YES;
        self.backgroundColor = [UIColor clearColor];

        self.backgroundView = [[UIView alloc] initWithFrame:self.bounds];
        self.backgroundView.backgroundColor = [UIColor colorWithRed:232.0/255.0 green:232.0/255.0 blue:232.0/255.0 alpha:1.0];
        self.backgroundView.autoresizingMask = UIViewAutoresizingFlexibleWidth;
        [self addSubview:self.backgroundView];

        NSBundle *frameworkBundle = [NSBundle bundleForClass:[self class]];

        UIImageView *playerBgImageView = [[UIImageView alloc] initWithImage:[[UIImage imageNamed:@"player_player_bg"
                                                                                        inBundle:frameworkBundle
                                                                   compatibleWithTraitCollection:nil]
                                                                             stretchableImageWithLeftCapWidth:5 topCapHeight:5]];
        playerBgImageView.frame = CGRectMake(_inset, _inset, _playerWidth - _inset * 2, _playerHeight - _inset * 2);
        playerBgImageView.autoresizingMask = UIViewAutoresizingFlexibleWidth;
        [self addSubview:playerBgImageView];

        self.playPauseButton = [UIButton buttonWithType:UIButtonTypeCustom];
        self.playPauseButton.autoresizesSubviews = YES;
        self.playPauseButton.imageView.contentMode = UIViewContentModeScaleToFill;
        [self.playPauseButton setImage:[UIImage imageNamed:@"player_play"
                                              inBundle:frameworkBundle
                         compatibleWithTraitCollection:nil]
                          forState:UIControlStateNormal];
        [self.playPauseButton setFrame:CGRectMake(_inset, _inset, _playerHeight - 2 * _inset, _playerHeight - 2 * _inset)];
        [self.playPauseButton addTarget:self action:@selector(playOrPause) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:self.playPauseButton];

        float originY = _playerHeight / 2.0 - 34 / 2;
        float originX = self.playPauseButton.frame.origin.x + self.playPauseButton.frame.size.width + _inset;
        CGRect sliderFrame = CGRectMake(originX, originY, frame.size.width - originX - _inset * 2, 34);
        _slider = [[UISlider alloc] initWithFrame:sliderFrame];
        _slider.autoresizingMask = UIViewAutoresizingFlexibleWidth;
        [_slider setMaximumTrackImage:[[UIImage imageNamed:@"player_progress_bg"
                                                  inBundle:frameworkBundle
                             compatibleWithTraitCollection:nil]
                                       resizableImageWithCapInsets:UIEdgeInsetsMake(0, 10, 0, 10)]
                             forState:UIControlStateNormal];
        [_slider setMinimumTrackImage:[[UIImage imageNamed:@"player_progress_blue"
                                                  inBundle:frameworkBundle
                             compatibleWithTraitCollection:nil]
                                       resizableImageWithCapInsets:UIEdgeInsetsMake(0, 10, 0, 10)]
                             forState:UIControlStateNormal];
        [_slider setThumbImage:[UIImage imageNamed:@"player_circle"
                                          inBundle:frameworkBundle
                     compatibleWithTraitCollection:nil]
                      forState:UIControlStateNormal];
        _slider.minimumValue = 0.0;
        _slider.maximumValue = self.audioPlayer.duration;
        [_slider addTarget:self action:@selector(sliderChanged:) forControlEvents:UIControlEventValueChanged];
        [self addSubview:_slider];

        self.bubbleView = [[UIView alloc] initWithFrame:CGRectMake(160, _slider.frame.origin.y - 46 + _slider.frame.size.height / 2, 72, 46)];
        self.bubbleView.backgroundColor = [UIColor clearColor];
        self.bubbleView.frame = [self createCurrentPositionFrame];
        UIImageView *bubbleImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"player_bubble"
                                                                                     inBundle:frameworkBundle
                                                                compatibleWithTraitCollection:nil]];
        bubbleImageView.contentMode = UIViewContentModeScaleToFill;
        bubbleImageView.frame = self.bubbleView.bounds;
        [self.bubbleView addSubview:bubbleImageView];
        self.timeLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 10, 72, 11)];
        self.timeLabel.backgroundColor = [UIColor clearColor];
        self.timeLabel.textColor = [UIColor blackColor];
        self.timeLabel.textAlignment = NSTextAlignmentCenter;
        self.timeLabel.font = [UIFont fontWithName:@"HelveticaNeue" size:10];
        self.timeLabel.text = [self calculateCurrentDuration];
        [self.bubbleView addSubview:self.timeLabel];

        self.bubbleView.hidden = YES;
        [self addSubview:self.bubbleView];

        self.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleTopMargin;

        self.duration = self.audioPlayer.duration;
    }

    return self;
}

- (void)initAudioPlayerWithData:(NSData *)audioData
{
    NSError *error = nil;
    self.audioPlayer = [[AVAudioPlayer alloc] initWithData:audioData error:&error];
    NSAssert(error == nil, @"Audio not found");
}

- (void)initAudioPlayerWithString:(NSString *)audioFilePath
{
    _audioFilePath = audioFilePath;
    NSAssert(audioFilePath != nil, @"Audio file path cannot be nil");

    NSURL *url = [NSURL fileURLWithPath:audioFilePath];
    NSError *error = nil;
    self.audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:url error:&error];
    NSAssert(error == nil, @"Audio file not found");
}

#pragma mark - Public methods

- (void)play
{
    if (! self.audioPlayer.isPlaying) {
        [self.audioPlayer play];
        self.isBubbleViewVisible = YES;
        [self updatePlayButtonImage];
    }
}

- (void)pause
{
    if (self.audioPlayer.isPlaying) {
        [self.audioPlayer pause];
        [self updatePlayButtonImage];
    }
}

- (void)dismiss
{
    self.audioPlayer = nil;
    [self removeFromSuperview];
}

- (void)setVolume:(float)volume
{
    self.audioPlayer.volume = volume;
}

#pragma mark - Private methods

- (void)playOrPause
{
    if (self.audioPlayer.isPlaying) {
        [self pause];
    }
    else {
        [self play];
    }
}

- (void)updatePlayButtonImage
{
    NSBundle *frameworkBundle = [NSBundle bundleForClass:[self class]];
    NSString *imageName = self.audioPlayer.isPlaying ? @"player_pause" : @"player_play";
    [self.playPauseButton setImage:[UIImage imageNamed:imageName
                                          inBundle:frameworkBundle
                     compatibleWithTraitCollection:nil]
                      forState:UIControlStateNormal];
}

- (void)onTimer:(NSTimer *)timer
{

    self.timeLabel.text = [self calculateCurrentDuration];

    [self.slider setValue:self.audioPlayer.currentTime animated:NO];
    self.bubbleView.frame = [self createCurrentPositionFrame];
}

- (float)xPositionFromSliderValue:(UISlider *)aSlider;
{
    float sliderRange = aSlider.frame.size.width - aSlider.currentThumbImage.size.width;
    float sliderOrigin = aSlider.frame.origin.x + (aSlider.currentThumbImage.size.width / 2.0);

    float sliderValueToPixels = ((aSlider.value-aSlider.minimumValue)/(aSlider.maximumValue-aSlider.minimumValue) * sliderRange) + sliderOrigin;

    return sliderValueToPixels;
}

- (void)sliderChanged:(UISlider *)slider
{
    [self.audioPlayer setCurrentTime:(int)slider.value];
}

- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{
    [self updatePlayButtonImage];
}

- (void)setIsBubbleViewVisible:(BOOL)isBubbleViewVisible
{
    _isBubbleViewVisible = isBubbleViewVisible;

    self.bubbleView.hidden = ! isBubbleViewVisible;
}

- (NSString *)calculateCurrentDuration
{
    long currentPlaybackTime = self.audioPlayer.currentTime;

    self.currentSecond = (int)currentPlaybackTime;
    
    int currentHours = (int)(currentPlaybackTime / 3600);
    int currentMinutes = (int)((currentPlaybackTime / 60) - currentHours*60);
    int currentSeconds = (int)(currentPlaybackTime % 60);
    NSString *currentTimeString = (currentHours > 0) ? [NSString stringWithFormat:@"%i:%02d:%02d", currentHours, currentMinutes, currentSeconds] : [NSString stringWithFormat:@"%02d:%02d", currentMinutes, currentSeconds];

    [self.delegate didUpdatePlaybackTimeAudioPlayer:self];
    
    return [NSString stringWithFormat:@"%@ / %@", currentTimeString, self.durationString];
}

- (CGRect)createCurrentPositionFrame
{
    CGRect frame = self.bubbleView.frame;
    frame.origin.x = [self xPositionFromSliderValue:self.slider] - self.bubbleView.frame.size.width / 2.0;
    return frame;
}

- (void)setBackgroundViewColor:(UIColor *)backgroundViewColor
{
    _backgroundViewColor = backgroundViewColor;

    self.backgroundView.backgroundColor = backgroundViewColor;
}

@end

File: DKAudioPlayer/Sources/DKAudioPlayer/DKAudioPlayer.swift
//
//  DKAudioPlayer.swift
//  DKAudioPlayer
//
//  Created by Denis Kutlubaev on 02.10.2022.
//

import AVFoundation
import Combine
import UIKit

protocol DKAudioPlayerDelegate: AnyObject {

    /// Called when audio player updates a playback time
    /// - Parameter audioPlayer: Instance of DKAudioPlayer
    func didUpdatePlaybackTime(in audioPlayer: DKAudioPlayer)
}

@objc
public class DKAudioPlayer: UIView {
    weak var delegate: DKAudioPlayerDelegate?

    var volume: Float = 0.5

    var duration: TimeInterval = 0.0

    var currentSecond = 0
    
    @objc
    public init(audioFilePath: String) {
        self.viewModel = DKAudioPlayerViewModel(audioFilePath: audioFilePath)

        super.init(frame: .zero)

        setupUI()
        setupLayout()
        bindViewModel()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    @objc
    public func play() {
        viewModel.play()
    }

    @objc
    public func dismiss() {
        viewModel.pause()
        removeFromSuperview()
    }

    private lazy var timeLabel: UILabel = {
        let label = UILabel(frame: .zero)
        label.translatesAutoresizingMaskIntoConstraints = false

        return label
    }()

    private lazy var bubbleView: UIView = {
        let view = UIView(frame: .zero)
        view.translatesAutoresizingMaskIntoConstraints = false
        view.isHidden = true

        return view
    }()

    private lazy var bubbleImageView: UIImageView = {
        let imageView = UIImageView(frame: .zero)
        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.image = Constants.bubbleImage
        imageView.contentMode = .scaleToFill

        return imageView
    }()

    private lazy var bubbleTimeLabel: UILabel = {
        let label = UILabel(frame: .zero)
        label.translatesAutoresizingMaskIntoConstraints = false
        label.backgroundColor = .clear
        label.textAlignment = .center
        label.font = Constants.bubbleTimeLabelFont
        label.text = "22:58"

        return label
    }()

    private lazy var slider: UISlider = {
        let slider = UISlider(frame: .zero)
        slider.translatesAutoresizingMaskIntoConstraints = false
        slider.setMaximumTrackImage(Constants.maximumTrackImage, for: .normal)
        slider.setMinimumTrackImage(Constants.minimumTrackImage, for: .normal)
        slider.setThumbImage(Constants.thumbImage, for: .normal)
        slider.minimumValue = 0.0
        slider.maximumValue = Float(viewModel.audioDuration)
        slider.isContinuous = false
        slider.addTarget(self, action: #selector(sliderValueChanged), for: .valueChanged)
        slider.addTarget(self, action: #selector(didBeginDraggingSlider), for: .touchDown)
        slider.addTarget(self, action: #selector(didEndDraggingSlider), for: .valueChanged)

        return slider
    }()

    private lazy var backgroundView: UIView = {
        let view = UIView(frame: .zero)
        view.translatesAutoresizingMaskIntoConstraints = false
        view.backgroundColor = Constants.backgroundColor

        return view
    }()

    private lazy var playerBgImageView: UIImageView = {
        let imageView = UIImageView(frame: .zero)
        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.image = Constants.backgroundImage
        imageView.contentMode = .scaleToFill
        imageView.clipsToBounds = true

        return imageView
    }()

    private lazy var playPauseButton: UIButton = {
        let button = UIButton(frame: .zero)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.setImage(Constants.playImage, for: .normal)
        button.addTarget(self, action: #selector(playOrPauseTapped), for: .touchUpInside)

        return button
    }()

    private lazy var bubbleViewLeadingConstraint: NSLayoutConstraint = {
        let constraint = bubbleView.leadingAnchor.constraint(equalTo: leadingAnchor)
        return constraint
    }()

    private var disposeBag = Set<AnyCancellable>()

    private let viewModel: DKAudioPlayerViewModel

}

// MARK: - Private

private extension DKAudioPlayer {

    static let bundle = Bundle(for: DKAudioPlayer.self)

    enum Constants {
        static let inset = 15 as CGFloat

        static let bubbleTimeLabelFont = UIFont.systemFont(ofSize: 9)

        // TODO: Move color to color assets or use system colors
        static let backgroundColor = UIColor(
            red: 232.0 / 255.0,
            green: 232.0 / 255.0,
            blue: 232.0 / 255.0,
            alpha: 1.0
        )

        static let leftCapWidth = 5
        static let topCapHeight = 5
        static let trackImageEdgeInsets = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)

        static let backgroundImage: UIImage? = {
            DKImage(named: "player_player_bg")?
                .stretchableImage(
                    withLeftCapWidth: Constants.leftCapWidth,
                    topCapHeight: Constants.topCapHeight
                )
        }()

        static let playImage: UIImage? = {
            DKImage(named: "player_play")
        }()

        static let pauseImage: UIImage? = {
            DKImage(named: "player_pause")
        }()

        static let maximumTrackImage: UIImage? = {
            DKImage(named: "player_progress_bg")?
                .resizableImage(
                    withCapInsets: Constants.trackImageEdgeInsets
                )
        }()

        static let minimumTrackImage: UIImage? = {
            DKImage(named: "player_progress_blue")?
                .resizableImage(
                    withCapInsets: Constants.trackImageEdgeInsets
                )
        }()

        static let thumbImage: UIImage? = {
            DKImage(named: "player_circle")
        }()

        static let bubbleImage: UIImage? = {
            DKImage(named: "player_bubble")
        }()
    }

    func setupUI() {
        translatesAutoresizingMaskIntoConstraints = false

        addSubview(playerBgImageView)
        addSubview(playPauseButton)
        addSubview(slider)
        addSubview(bubbleView)
        bubbleView.addSubview(bubbleImageView)
        bubbleView.addSubview(bubbleTimeLabel)
    }

    func setupLayout() {
        setupPlayerBgImageView()
        setupPlayPauseButton()
        setupSlider()
        setupBubbleView()
        setupBubbleImageView()
        setupBubbleTimeLabel()
    }

    func setupPlayerBgImageView() {
        addConstraints([
            playerBgImageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constants.inset),
            playerBgImageView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -Constants.inset),
            playerBgImageView.topAnchor.constraint(equalTo: topAnchor, constant: Constants.inset),
            playerBgImageView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -Constants.inset)
        ])
    }

    func setupPlayPauseButton() {
        addConstraints([
            playPauseButton.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constants.inset),
            playPauseButton.topAnchor.constraint(equalTo: playerBgImageView.topAnchor),
            playPauseButton.bottomAnchor.constraint(equalTo: playerBgImageView.bottomAnchor),
            playPauseButton.widthAnchor.constraint(equalToConstant: 44)
        ])
    }

    func setupSlider() {
        addConstraints([
            slider.leadingAnchor.constraint(equalTo: playPauseButton.trailingAnchor),
            slider.trailingAnchor.constraint(equalTo: playerBgImageView.trailingAnchor, constant: -Constants.inset),
            slider.centerYAnchor.constraint(equalTo: centerYAnchor)
        ])
    }

    func setupBubbleView() {
        addConstraints([
            bubbleView.bottomAnchor.constraint(equalTo: slider.topAnchor),
            bubbleViewLeadingConstraint
        ])
    }

    func setupBubbleImageView() {
        bubbleView.addConstraints([
            bubbleImageView.leadingAnchor.constraint(equalTo: bubbleView.leadingAnchor),
            bubbleImageView.trailingAnchor.constraint(equalTo: bubbleView.trailingAnchor),
            bubbleImageView.topAnchor.constraint(equalTo: bubbleView.topAnchor),
            bubbleImageView.bottomAnchor.constraint(equalTo: bubbleView.bottomAnchor)
        ])
    }

    func setupBubbleTimeLabel() {
        bubbleView.addConstraints([
            bubbleTimeLabel.leadingAnchor.constraint(equalTo: bubbleView.leadingAnchor, constant: 6),
            bubbleTimeLabel.trailingAnchor.constraint(equalTo: bubbleView.trailingAnchor, constant: -6),
            bubbleTimeLabel.topAnchor.constraint(equalTo: bubbleView.topAnchor, constant: 7)
        ])
    }

    @objc
    func playOrPauseTapped(_ sender: UIButton) {
        viewModel.playOrPause()
    }

    @objc
    func sliderValueChanged(_ sender: UISlider) {
        viewModel.updateCurrentTime(to: TimeInterval(sender.value))
    }

    func bindViewModel() {
        viewModel.isPlaying.sink { [unowned self] isPlaying in
            let image = isPlaying ? Constants.pauseImage : Constants.playImage
            self.playPauseButton.setImage(image, for: .normal)
        }
        .store(in: &disposeBag)

        viewModel.currentTime.sink { [unowned self] currentTime in
            self.slider.setValue(Float(currentTime), animated: false)

            if self.bubbleView.bounds.size.width > 0 {
                self.bubbleViewLeadingConstraint.constant = xPositionFromSliderValue(self.slider) - self.bubbleView.bounds.size.width / 2
                self.bubbleView.setNeedsLayout()
                self.bubbleView.layoutIfNeeded()
                self.bubbleView.isHidden = false
            } else {
                self.bubbleView.isHidden = true
            }
        }
        .store(in: &disposeBag)

        viewModel.currentTimeText.sink { [unowned self] currentTimeText in
            self.bubbleTimeLabel.text = currentTimeText
            self.bubbleView.isHidden = currentTimeText.isEmpty || self.bubbleView.bounds.size.width == 0
        }
        .store(in: &disposeBag)
    }

    func xPositionFromSliderValue(_ slider: UISlider) -> CGFloat {
        guard let currentThumbImage = slider.currentThumbImage else { return 0 }

        let sliderRange = slider.frame.size.width - currentThumbImage.size.width
        let sliderOrigin = slider.frame.origin.x + currentThumbImage.size.width / 2.0
        let sliderValueToPixels = (CGFloat((slider.value - slider.minimumValue) / (slider.maximumValue - slider.minimumValue)) * sliderRange) + sliderOrigin

        return CGFloat(sliderValueToPixels)
    }

    @objc
    func didBeginDraggingSlider() {
        bubbleView.isHidden = true
        viewModel.pauseUpdates()
    }

    @objc
    func didEndDraggingSlider() {
        bubbleViewLeadingConstraint.constant = xPositionFromSliderValue(self.slider) - self.bubbleView.bounds.size.width / 2
        bubbleView.setNeedsLayout()
        bubbleView.layoutIfNeeded()
        bubbleView.isHidden = false
        viewModel.resumeUpdates()
    }
    
}
